src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,144 @@
+                    List<PhysicalPlanWithStatistics> tmpCandidateSubPlans = getSubPlansContainVars( currentJoinPlan.getExpectedVariables(), subPlansWithStatistics );
+                    if( tmpCandidateSubPlans.isEmpty() ){
+                        tmpCandidateSubPlans = subPlansWithStatistics;
+                    }
+
+                    // Keep candidate SubPlans that use the same vocabulary mapping with the previous selected SubPlan.
+                    // If 'candidateSubPlans' is empty, use all subplans in 'tmpCandidateSubPlans' as candidate subplans.
+                    List<PhysicalPlanWithStatistics> candidateSubPlans = getSubPlansWithSameVoc( nextSelectedSubPlan.getVocabularyMappings(), tmpCandidateSubPlans);
+                    if( candidateSubPlans.isEmpty() ){
+                        candidateSubPlans = tmpCandidateSubPlans;
+                    }
+
+                    // The candidateSubPlans are added to the ordered list 'orderedCandidateSubPlans'
+                    orderedCandidateSubPlans.addAll(candidateSubPlans);
+
+                    // select the first element (with the lowest cardinality) from orderedCandidateSubPlans as the nextSelectedSubPlan
+                    // Remove the selected subPlan from subPlansWithStatistics and clear orderedCandidateSubPlans for the next loop
+                    nextSelectedSubPlan = orderedCandidateSubPlans.poll();
+                    subPlansWithStatistics.remove(nextSelectedSubPlan);
+                    orderedCandidateSubPlans.clear();
+
+                    // Construct a physical plan using bind join
+                    currentJoinPlan = constructPlanWithBJ( currentJoinPlan, nextSelectedSubPlan.plan );
+                }
+
+                return currentJoinPlan;
+            }
+        };
+    }
+
+    /**
+     * Iterate through the remaining subplans and selected those that have the same vocabulary mapping in the given list of vocabulary mappings.
+     */
+    protected List<PhysicalPlanWithStatistics> getSubPlansWithSameVoc( final List<VocabularyMapping> vms, final List<PhysicalPlanWithStatistics> subPlansWithStatistics ) {
+        final List<PhysicalPlanWithStatistics> subPlansWithSameVoc = new ArrayList<>();
+        for ( final PhysicalPlanWithStatistics subplan: subPlansWithStatistics ) {
+            final List<VocabularyMapping> vmsNext = subplan.getVocabularyMappings();
+            vms.retainAll(vmsNext);
+            if ( !vms.isEmpty() ){
+                subPlansWithSameVoc.add(subplan);
+            }
+        }
+
+        return subPlansWithSameVoc;
+    }
+
+    /**
+     * Construct a physical plan using bind join when it is possible:
+     * If both currentPlan and nextPlan have PhysicalOpLocalToGlobal as root operator, and use the same vocabulary mapping,
+     * the physical plan can be simplified as when unary operator can be applied:
+     * tpAdd^{tp}_{fm}(g2l^{vm}(l2g^{vm}(p))) --> tpAdd^{tp}_{fm}(p)
```

Comment:
While this equivalence is correct, the function is not just doing that but, instead, it is actually converting a binary join into a tpAdd. I mean, in contrast to the formula given here, there is no g2l operator anywhere in the two arguments (`currentPlan` and `nextPlan`).

Since the function is actually converting a binary join into a tpAdd, it is not correct in general. Instead, it has still the same issue as mentioned in the comments of my previous review for this PR. In the same way, the equivalences that you have in Proposition 1 of the paper draft are also not correct in general. They only hold for specific vocabulary mappings.

Reply:
I take back what I said about the equivalences in the paper draft. Given that they are only for triple patterns that cannot produce solution mappings with IRIs of classes or properties, they are correct. However, I would like both the equivalences and the implementation to be less restrictive. In particular, the equivalences can also be used for vocab.mappings that do not use subclass-based rules, in which case the patterns of the query can be anything. Should I put what I mean into the paper draft?

Reply:
Hi Olaf, I would appreciate it if you could put your intended meaning in the paper draft. 
Thanks.

Reply:
Done now. Sorry for the delay! Take a look and email me if you have questions or comments about it.

Reply:
Hi Olaf, I sincerely appreciate you taking time out of your tight schedule to add text. The text described is very clear, and I have only a minor question. 
Specifically, I am uncertain if the first equivalence of Proposition 4 is necessary because it can be further simplified to join(\varphi, \varphi'). In other words, it appears to imply the same rule as the second proposition.

Regarding the implementation of plan rewriting, I have not found a good way to verify if every rule in the vocabulary mapping is either an equivalence on concepts or an equivalence on roles. Do you have any suggestions?

Reply:
> I am uncertain if the first equivalence of Proposition 4 is necessary because it can be further simplified to join(\varphi, \varphi'). In other words, it appears to imply the same rule as the second proposition.

Indeed, that equivalence is redundant because of the second equivalence. Great observation! I have removed it.

> I have not found a good way to verify if every rule in the vocabulary mapping is either an equivalence on concepts or an equivalence on roles.

The issue is in the `VocabularyMappingImpl` class, right? I suggest that you extend the `VocabularyMapping` interface with a `boolean`-typed function called `isEquivalenceOnly()`. Then, in the `VocabularyMappingImpl` class, you implement this function by checking whether the `Graph vocabularyMapping` does not contain triples with predicates rdfs:subClassOf, rdfs:subPropertyOf, owl:inverseOf, owl:unionOf, and owl:intersectionOf. Use the corresponding Java classes for the RDFS and the OWL vocabulary (as done in other parts of the file).